src/differential_expression/deseq2/script.R

```diff
@@ -0,0 +1,458 @@
+    )
+
+    # Convert to data frame and add metadata
+    results_df <- as.data.frame(res)
+    results_df$gene_id <- rownames(results_df)
+    results_df$contrast <- paste0(contrast_spec[2], "_vs_", contrast_spec[3])
+    results_df$comparison_group <- contrast_spec[2]
+    results_df$control_group <- contrast_spec[3]
+    results_df$abs_log2FoldChange <- abs(results_df$log2FoldChange)
+    results_df$significant <- (
+      results_df$padj < par$p_adj_threshold &
+        !is.na(results_df$padj) &
+        abs(results_df$log2FoldChange) > par$log2fc_threshold
+    )
+
+    # Sort by effect size
+    results_df[order(results_df$log2FoldChange, decreasing = TRUE), ]
+  })
+
+  # Combine all results
+  combined_results <- do.call(rbind, all_results)
+
+  # Log summary statistics
+  for (i in seq_along(contrast_specs)) {
+    contrast_spec <- contrast_specs[[i]]
+    contrast_name <- paste0(contrast_spec[2], "_vs_", contrast_spec[3])
+    contrast_subset <- combined_results[
+      combined_results$contrast == contrast_name,
+    ]
+    n_significant <- sum(contrast_subset$significant, na.rm = TRUE)
+    cat("Contrast", contrast_name, ":", n_significant, "significant genes\n")
+  }
+
+  combined_results
+}
+
+# Save results and print summary statistics
+save_results_and_log_summary <- function(
+    results, output_file, cell_group = NULL) {
+  group_text <- if (!is.null(cell_group)) paste(" for", cell_group) else ""
+  cat("Saving results", group_text, "to", output_file, "\n")
+
+  write.csv(results, output_file, row.names = FALSE)
+
+  # Calculate summary statistics
+  sig_results <- results[results$significant & !is.na(results$significant), ]
+  upregulated <- sig_results[sig_results$log2FoldChange > 0, ]
+  downregulated <- sig_results[sig_results$log2FoldChange < 0, ]
+
+  cat("Summary", group_text, ":\n")
+  cat("  Total genes analyzed:", nrow(results), "\n")
+  cat("  Significant upregulated:", nrow(upregulated), "\n")
+  cat("  Significant downregulated:", nrow(downregulated), "\n")
+}
+
+# Main analysis workflow
+main <- function() {
+  cat("Loading pseudobulk data from", par$input, "\n")
+
+  # Load and prepare data
+  h5ad_path <- h5mu_to_h5ad(par$input, par$modality)
+  mod <- anndataR::read_h5ad(h5ad_path, as = "InMemoryAnnData")
+  metadata <- as.data.frame(mod$obs)
+
+  # Get expression matrix
+  layer <- if (!is.null(par$input_layer)) {
+    mod$layers[[par$input_layer]]
+  } else {
+    mod$X
+  }
+
+  if (is_normalized(layer)) {
+    stop("Input layer must contain raw counts.")
+  }
+
+  # Prepare analysis components
+  cat("Preparing design formula\n")
+  design_factors <- parse_design_formula(par$design_formula)
+
+  cat("Preparing contrast matrix\n")
+  contrast_specs <- prepare_contrast_matrix(
+    design_factors, par$contrast_column, metadata
+  )
+
+  cat("Preparing counts matrix for DESeq2\n")
+  var_names <- if (
+    !is.null(par$var_gene_name) && par$var_gene_name %in% colnames(mod$var)
```

Comment:
I think an error should be provided when `var_gene_name` is specified but not present in the colnames